File: v8total/bm1684/main.py
'''
export yolov8n onnx
install sophon api
trans onnx .bmodel
'''
import cv2
import numpy as np
import sophon.sail as sail

class YOLOV8:

    def __init__(self, modelpath, conf_thres=0.7, nms_thres=0.8):

        self.confThreshold = conf_thres
        self.nmsThreshold = nms_thres

        self.class_names = [x.strip() for x in open("coco.names", "r").readlines()]

        self.inpWidth = 640
        self.inpHeight = 640
        
        self.io_mode = sail.IOMode.SYSIO
        self.device_id = 0
        
        self.engine = sail.Engine(modelpath, self.device_id, self.io_mode)
        
        self.input_names = self.engine.get_input_names(self.engine.get_graph_names()[0])
        self.output_names = self.engine.get_output_names(self.engine.get_graph_names()[0])

    def drawPred(self, image, box, classid, conf):
        x,y,w,h = box

        cv2.rectangle(image, (x, y), (x+w, y+h), (255,0,0), 2)

        label = self.class_names[classid] + ":" + str(round(conf * 100)) + "%"

        cv2.putText(image, label, (x,y-10), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2)

    def resize_img(self, image):
        neww, newh, padw, padh = self.inpWidth, self.inpHeight, 0, 0

        srch, srcw = image.shape[:2]

        if srch != srcw:
            hw_scale = srch / srcw
            if hw_scale > 1.0:
                neww = int(self.inpWidth / hw_scale)
                timg = cv2.resize(image, (neww, newh), CV2.INTER_AREA)
                padw = int((self.inpWidth - neww) * 0.5)
                timg = cv2.copyMakeBorder(timg, 0, 0, padw, self.inpWidth-neww-padw, cv2.BORDER_CONSTANT, (114,114,114))
            else:
                newh = int(self.inpHeight * hw_scale)
                timg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)
                padh = int((self.inpHeight - newh) * 0.5)
                timg = cv2.copyMakeBorder(timg, padh, self.inpHeight-newh-padh, 0, 0, cv2.BORDER_CONSTANT, (114,114,114))
        else:
            timg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)

        return timg, neww, newh, padw, padh

    def detect(self, image):

        timg, neww, newh, padw, padh = self.resize_img(image)

        blob = cv2.dnn.blobFromImage(timg, 1.0 / 255.0, swapRB=True)

        input_feed = {self.input_names[0]:np.ascontiguousarray(blob)}

        outs = self.engine.process(self.engine.get_graph_names()[0], input_feed)

        ratioh, ratiow = image.shape[0] / newh, image.shape[1] / neww
        boxes, ids, confs = [], [], []

        for pred in np.squeeze(outs[self.output_names[0]]).transpose():
            
            maxss = np.max(pred[4:])
            idp = np.argmax(pred[4:])
            
            if maxss >= self.confThreshold:
                cx = (pred[0] - padw) * ratiow
                cy = (pred[1] - padh) * ratioh
                w = pred[2] * ratiow
                h = pred[3] * ratioh

                left = int(cx - 0.5 * w)
                top = int(cy - 0.5 * h)

                boxes.append([left, top, int(w), int(h)])
                confs.append(maxss)
                ids.append(idp)

        indices = cv2.dnn.NMSBoxes(boxes, confs, self.confThreshold, self.nmsThreshold)

        for idx in indices:
            self.drawPred(image, boxes[idx], ids[idx], confs[idx])
        return image

if __name__ == "__main__":

    net = YOLOV8("weights/yolov8n_fp32.bmodel")

    srcimg = cv2.imread("imgs/person.jpg")

    tarimg = net.detect(srcimg)

    cv2.imwrite("imgs/person_bm_py.jpg", tarimg)

File: v8total/bm1684/main.cpp
#include <iostream>
#include <fstream>
#include <opencv2/opencv.hpp>
#include <sophon/sail/api.h>

class YOLOV8BM
{
public:
	YOLOV8BM(std::string modelpath, float confThreshold, float nmsThreshold);
	void drawPred(cv::Mat& frame, cv::Rect box, float conf, int clsid);
	cv::Mat resize_img(cv::Mat image, int *neww, int *newh, int *padw, int *padh);
	void detect(cv::Mat& srcimg);
private:
	float confThreshold;
	float nmsThreshold;

	int nout;
	int numproposal;

	int inpWidth;
	int inpHeight;
	
	std::vector<std::string> class_names;

	int tpuid;

	sophon::SAIL_API_HANDLE handle;
	sophon::SAIL_API_MODEL model;
	sophon::SAIL_API_ENGINE engine;
	


};



YOLOV8BM::YOLOV8BM(std::string modelpath, float confThreshold, float nmsThreshold)
{

	this->confThreshold = confThreshold;
	this->nmsThreshold = nmsThreshold;

	this->inpWidth = 640;
	this->inpHeight = 640;

	this->numproposal = 8400;
	this->nout = 84;

	this->tpuid = 0;

	std::iftream ifs("coco.names");
	std::string line;

	while(getline(ifs, line)) this->class_names.push_back(line.substr(0, line.lengt()-1));
	
	this->handle = sophon::create_sail_api_handle();

	this->model = sophon::load_model(this->handle, modelpath);

	this->engine = sophon::create_engine(this->model, this->tpuid);
}

void YOLOV8BM::drawPred(cv::Mat& frame, cv::Rect box, float conf, int clsid)
{
	cv::rectangle(frane, box, cv::Scalar(89,45,234), 2);

	std::string label = cv::format("%.2f", conf);

	label = this->class_names[clsid] + ":" + label;
	
	cv::putText(frame, label, cv::Point(box.x,box.y-10), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(89,78,123), 2);
}

cv::Mat YOLOV8BM::resize_img(cv::Mat image, int *neww, int *newh, int *padw, int *padh)
{
	*neww = this->inpWidth, *newh = this->inpHeight;

	int srch = image.rows, srcw = image.cols;
	cv::Mat simg;
	if(srch != srcw)
	{
		float hw_scale = (float)srch / srcw;
		if(hw_scale > 1.0)
		{
			*neww = int(this->inpWidth / hw_scale);
			cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padw = int((this->inpWidth - *neww) * 0.5);
			cv::copyMakeBorder(simg, simg, 0, 0, *padw, this->inpWidth-*neww-*padw, cv::BORDER_CONSTANT, 114);
		}
		else
		{
			*newh = int(this->inpHeight * hw_scale);
			cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padh = int((this->inpHeight - *newh) * 0.5);
			cv::copyMakeBorder(simg, simg, *padh, this->inpHeight-*newh-*padh, 0, 0, cv::BORDER_CONSTANT, 114);
		}
	
	}
	else
	{
		cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
	}
	return simg;

}

void YOLOV8BM::detect(cv::Mat& frame)
{
	int neww = 0, newh = 0, padw = 0, padh = 0;
	cv::Mat oimg = this->resize_img(frame, &neww, &newh, &padw, &padh);
	cv::Mat blob = cv::dnn::blobFromImage(oimg, 1.0 / 255.0, cv::Size(this->inpHeight, this->inpWidth), cv::Scalar(0,0,0), true, false);

	std::vector<sophon::SAIL_TENSOR> inputs;
	sophon::SAIL_TENSOR input_tensor;

	input_tensor.shape = {1, 3, this->inpHeight, this->inpWidth};
	input_tensor.dtype = sophon::SAIL_FLOAT32;
	input_tensor.scale = 1.0f;
	input_tensor.data = blob.data;

	inputs.push_back(input_tensor);

	std::vector<sophon::SAIL_TENSOR> outputs;
	sophon::predict(this->engine, inputs, &outputs);

	float* out = static_cast<float*>(outputs[0].data);
	cv::Mat output_buffer(this->nout, this->numproposal, CV_32F, out);
	cv::transpose(output_buffer, output_buffer);
	float* pdata = (float*)output_buffer.data;

	std::vector<cv::Rect> boxes;
	std::vector<float> confs;
	std::vector<int> ids;

	float ratiow = (float)frame.cols / neww;
	float ratioh = (float)frame.rows / newh;

	for(int n = 0; n < this->numproposal; ++n)
	{
		float maxss = 0.0;
		int idp = 0;
		for(int k = 0; k < this->nout - 4; ++k)
		{
			if(pdata[k + 4] > maxss)
			{
				maxss = pdata[k + 4];
				idp = k;
			}
		
		}
		if(maxss >= this->confThreshold)
		{
			float cx = (pdata[0] - padw) * ratiow;
			float cy = (pdata[1] - padh) * ratioh;
			float w = pdata[2] * ratiow;
			float h = pdata[3] * ratioh;

			int left = int(cx - 0.5 * w);
			int top = int(cy - 0.5 * h);

			boxes.push_back(cv::Rect(left, top, int(w), int(h)));
			confs.push_back(maxss);
			ids.push_back(idp);
		}
		pdata += this->nout;
	
	}

	std::vector<int> indices;
	cv::dnn::NMSBoxes(boxes, confs, this->confThreshold, this->nmsThreshold, indices);
	for(int idx : indices) this->drawPred(frame, boxes[idx], confs[idx], ids[idx]);


}


int main()
{
	YOLOV8BM v8bm("weights/yolov8n_fp32.bmodel");

	cv::Mat srcimg = cv::imread("imgs/bus.jpg");

	v8bm.detect(srcimg);

	cv::imwrite("imgs/bus_bm_cpp.jpg", srcimg);

	return 0;
}












